woodwork/type_sys/inference_functions.py

```diff
@@ -100,8 +106,13 @@ def double_func(series):
     return False
 
 
-def boolean_func(series):
-    if boolean_nullable_func(series) and not series.isnull().any():
+def boolean_func(series, is_boolean_nullable=None):
```

Comment:
Admittedly, adding an argument that defaults to None is probably not where we want to land long-term, but I think this is a fair short-term fix. I think a better long term solution would be to generalize the idea of caching in the `get_inference_matches` function itself.